### PR TITLE
Clean up upgrade.php

### PR DESCRIPTION
**Pull request by NanoSector**
When running upgrade.php I noticed a lot of issues pop up. I'm going to clean it up as much as I can to hopefully make maintenance easier.

### CODE REVIEW

**Comment by NanoSector** [comment]
Do we really need CleanupMods? It seems to be touching a whole lot of stuff the upgrader shouldn't really be doing (it parses a modification's XML file to see if it's still valid, something the package manager in the admin panel does just fine). If you've just upgraded you really shouldn't be touching old mods.

**Comment by albertlast** [comment]
When your already touching the upgrade.php,
an "step" who is missing the config rewrite from mysql to mysqli see the part of autoswitch: https://github.com/SimpleMachines/SMF2.1/issues/3583

**Comment by jdarwood007** [comment]
CleanupMods was added in 2.0 beta phase as a way to attempt to automatically reinstall mods.  It caused a lot of problems, so it was just disabled until we could get around to resolving those issues.  It never made it by the time 2.0 was gold.

**Comment by NanoSector** [comment]
We probably should look into recreating it in a sane way, then. A lot has changed and we should be able to use the Subs files' functions now instead of manually checking packages.

**Comment by sbulen** [comment]
Thanks Yoshi...   While in Upgrade.php, you might want to take a peek at issue #3554

**Comment by NanoSector** [comment]
I'll take a look, thanks.

**Comment by NanoSector** [comment]
Why do we populate the entire user array during upgrade? Shouldn't it be enough to just check if the user has admin access?

**Comment by live627** [comment]
The upgrader is still a complete mystery to me. Many thanks for fixing it.

**Comment by NanoSector** [comment]
I kind of want to avoid a last-minute rewrite to further stretch the next release but I really feel that a new upgrade/installation framework should be put in place. It is a big mess indeed, lots of old code.

The thing that worries me most about the current upgrader is that there is a lot of code copied from the Sources files. IMO, this is a direct consequence of having huge functions in many of those files. For instance, checkLogin copies huge chunks of password checking etc. from LogInOut.php, but it cannot use the functions inside LogInOut.php because they were written for a specific use case (and can't be used for any other use case). This in the end means that if we ever decide to change the behavior of those functions (or even worse, fix a security issue), Upgrade.php will not reflect these changes, possibly leaving vulnerabilities open. Fixing this then gets messy because some chunks are copied directly from the Sources files, some are altered a bit, so we're essentially forced to fix the same thing twice.

Fixing this mess is possible by splitting code like LogInOut.php into smaller, specialized functions; Upgrade.php can then simply directly include LogInOut.php and utilize these functions. This avoids code duplication as the same code can be reused internally too and allows easier maintenance of code (and as a bonus gets us a much higher code rating in Scrutinizer).

I'm not sure if this should be done in 2.1 though as it pretty much means rewriting half the codebase, but it is something I would love to see in 3.0.

**Comment by Oldiesmann** [comment]
At this point I think it's best to just clean up upgrade.php as much as we can and look at rewriting everything for the next version of SMF. It would be a lot of work to completely rewrite all of that at this point, especially given how long 2.1 has been in development.

**Comment by NanoSector** [comment]
Yeah. I'll probably get some more cleaning up done today.

**Comment by NanoSector** [comment]
When digging through the code I noticed we only recommend to backup your tables when upgrading from YaBB SE (and I changed that to always recommend it). That got me thinking though, do we really want to have the upgrade script backup tables? I mean, it's nice to have it, but shouldn't we encourage the user to make a full backup of their database first, since that's better than just copying tables?

File: other/upgrade.php [hunk 1]
```diff
@@ -28,15 +35,6 @@
...
-		'version' => '5.0.3',
-		'version_check' => 'return min(mysql_get_server_info(), mysql_get_client_info());',
-		'utf8_support' => true,
-		'utf8_version' => '5.0.3',
-		'utf8_version_check' => 'return mysql_get_server_info();',
-		'alter_support' => true,
-	),
 	'postgresql' => array(
 		'name' => 'PostgreSQL',
 		'version' => '8.0',
```
Comment (tinoest) [inline other/upgrade.php]: From this comment here; http://www.simplemachines.org/community/index.php?topic=547286.msg3890058#msg3890058

PostgreSQL is meant to be version 9.1 minimum now.

**Comment by jdarwood007** [comment]
Yabb != Yabb SE.
We still support upgrading from Yabb SE (see the upgrade_1.0.sql).  Yabb needs a converter.


Have we tested a SMF 1.1 or 2.0 install upgrading to 2.1 with mysqli?  Does the upgrader handler that smoothly?

**Comment by NanoSector** [comment]
But we have converters for both of them:
http://www.simplemachines.org/community/index.php?topic=140776.0

I would test but I only have php7 on my systems, that doesn't work with
smf2.0 or 1.1.

Op wo 4 jan. 2017 01:45 schreef Jeremy D <notifications@github.com>:

**Comment by NanoSector** [comment]
Sorry, didn't notice your comment; I'll fix that tomorrow.

Op di 3 jan. 2017 21:28 schreef Martyn Brown <notifications@github.com>:

**Comment by live627** [comment]
@Yoshi2889 do you wish to add more changes to this? Can I merge this?

**Comment by sbulen** [comment]
I'm looking forward to this set of changes!

**Comment by NanoSector** [comment]
I don't think I'd want to add more changes, thanks for fixing that postgre
issue. I cannot test the whole thing since I only have access to PHP 7 now
so if someone can test if it still upgrades fine from both 1.1 and 2.0 I
think it's good to go.

Op zo 8 jan. 2017 06:11 schreef sbulen <notifications@github.com>:

File: other/install.php [hunk 2]
```diff
@@ -523,9 +523,8 @@ function CheckFilesWritable()
 		'Themes',
 		'agreement.txt',
 		'Settings.php'
```
Comment (Oldiesmann) [inline other/install.php]: You're missing a comma at the end of line 525

**Comment by NanoSector** [comment]
Didnt we change the mysqli think back to mysql? I think Upgrade.php has not been changed accordingly, then..

**Comment by live627** [comment]
I already changed it.

**Comment by NanoSector** [comment]
Ah, cool.

**Pull request merged**